Add Per-Monitor DPI awareness to more controls

MaartenBent opened the pull request:
Add Per-Monitor DPI awareness to `wxDC` and `wxGraphicsContext` (font only), `wxSTC`, `wxPropertyGrid`, and `wxMenuItem`.

This is the last batch of changes from #334. As a nice side effect, font scaling in `wxGDIPlusPrintingContext` is not needed anymore.

Please don't merge yet, I want to do more testing. I created the PR to get some feedback.

MaartenBent committed:
Adjust font of wxDC to DPI of associated window

MaartenBent committed:
Support DPI change in wxStyledTextCtrl

Change to zoom-level and margin width based on the active DPI.
Send wxDPIChangedEvent to all controls and windows.

MaartenBent committed:
Fix wxPropertyGrid row height on DPI change

Fix collapse button size of wxPropertyGrid in High DPI.

MaartenBent committed:
Allow to retrieve the wxWindow associated with a wxBufferedPaintDC

When a wxBufferedPaintDC is created, the base classes does not initialize the
m_window variable with the used wxWindow. Only the associated m_paintdc
initializes this variable. Add a protected function that allows to set the
wxWindow of a wxDC so GetWindow() will return the window.

This fixes the font size of custom attributes in wxDataViewCtrl when the DPI
changes.

MaartenBent committed:
Fix font size of wxMenuItem with custom font on DPI change

Recalculate MenuDrawData sizes after the DPI changes, and reset the maximum
accelerator width.

MaartenBent committed:
Use associated window in wxMSWDCImpl::GetPPI()

MaartenBent committed:
Correctly use pixelSize in wxGraphicsContext::CreateFont

It is documented as using pixel size, and this is also used in all other contexts.
For clarity, rename all parameters to sizeInPixels.

This partially reverts d5020362ffb99608eacd2492d621cad20b1075a7

MaartenBent committed:
Add wxGraphicsRenderer::CreateFontAtDPI to support font with fractional pixel-size

vadz commented on src/msw/dc.cpp:
```diff
@@ -1439,6 +1439,9 @@ void wxMSWDCImpl::DoDrawRotatedText(const wxString& text,
     wxFont font = m_font.IsOk() ? m_font : *wxSWISS_FONT;
+    if ( !m_font.IsOk() && m_window )
+        font.WXAdjustToPPI(m_window->GetDPI());
+
```
This would result in a slightly larger diff, but I think it would be clearer to write this as
```cpp
wxFont font;
if ( m_font.IsOk() )
{
    font = m_font;
}
else // Use default font appropriate for rotated text.
{
    font = *wxSWISS_FONT;
    if ( m_window )
        font.WXAdjustToPPI(m_window->GetDPI());
}
```

BTW, I wonder if the comment about `DEFAULT_GUI_FONT` not being true type is still actually relevant (it was written in 4770df95a0a back in 2001 and certainly was true back then, but I must have tested it under something like Windows 2000...).

vadz commented on src/stc/stc.cpp:
```diff
@@ -239,6 +240,16 @@ bool wxStyledTextCtrl::Create(wxWindow *parent,
+    double activeDPI = parent->GetContentScaleFactor();
+
+    int ptSizeOld = StyleGetSize(wxSTC_STYLE_DEFAULT);
+    int ptSizeNew = (int)(ptSizeOld * activeDPI / baseDPI);
```
Should arguably use `::MulDiv(ptSizeOld, activeDPI, baseDPI)`.

vadz commented on include/wx/dc.h:
```diff
@@ -1364,6 +1366,9 @@ class WXDLLIMPEXP_CORE wxDC : public wxObject
 
     wxDCImpl * const m_pimpl;
 
+    void SetWindow(wxWindow* w)
```
An alternative could be to get rid of `wxDCImpl::m_window` entirely and make `GetWindow()` virtual to allow reimplementing it properly in the derived classes (the base class version would just return `NULL`), including `wxBufferedPaintDC` that would return `m_paintdc.GetWindow()` from it.

I think this would be slightly more proper, but I could be missing some compatibility implications and it would, of course, be more work...

vadz commented on src/common/dcgraph.cpp:
```diff
@@ -518,7 +518,7 @@ void wxGCDCImpl::SetFont( const wxFont &font )
     if ( m_graphicContext )
     {
-        m_graphicContext->SetFont(font, m_textForegroundColour);
+        m_graphicContext->SetFont(m_font, m_textForegroundColour);
```
Is there some deeper meaning to this change I'm missing?

vadz commented on include/wx/fontutil.h:
```diff
@@ -144,6 +144,9 @@ class WXDLLIMPEXP_CORE wxNativeFontInfo
     float        pointSize;
+    // Keep track of the ppi that is used to calculate the logfont height from
+    // the pointSize.
+    int          ppi;
```
IIRC we tried to avoid storing the PPI in `wxNativeFontInfo` itself initially. It's strange that we managed to get everything working without doing this except for, apparently, wxGC stuff. What's so different about it that it requires doing this?

vadz reviewed (commented):
As usual, thanks a lot for untangling (and finishing to untangle) all these changes!

I don't have any real comments, I think we can live with PPI in `wxNativeFontInfo`, even if I'd still prefer to avoid it if reasonably possible. But if not, let's merge this and rejoice.

Thanks again!

NewPagodi commented:
For the `wxStyledTextCtrl` files, it's necessary to modify stc.h.in and stc.cpp.in (and then rerun gen_iface.py).

MaartenBent commented:
> IIRC we tried to avoid storing the PPI in `wxNativeFontInfo` itself initially. It's strange that we managed to get everything working without doing this except for, apparently, wxGC stuff. What's so different about it that it requires doing this?

The difference is that both `wxGraphicsContext` supports fractional sizes. I can use `font.GetPixelSize().GetHeight()` instead, but then precision is lost. This is already visible with normal dpi in the drawing sample. On the text-page I use pointSize 5.5 and zoom 10x. With my change letters are 2 pixels higher than without.

To convert the fractional point size to the fractional pixel size, the active DPI is needed. It is easiest to store this with the font, instead of calling `*Renderer::CreateFont` and `wxD2DFontData/wxGDIPlusFontData` with extra parameters.

vadz commented:
Thinking more about it, I think I understand why did it like this now. To begin with, I definitely agree that we should use floating point size in `wxGC`. The problem is that while we've been able to adjust `LOGFONT::lfHeight` in `wxFont::WXAdjustToPPI()`, we're not doing anything to `pointSize` there and the reason we don't is, AFAICS, that we must return the originally specified size from `wxFont::GetFractionalPointSize()`. So we could do the same thing as we did for `wxDC`, but this would basically require adding another `float adjustedPointSize`, which is not better at all than just storing PPI as this PR does.

OTOH I wonder what's wrong with storing PPI in `wx{D2D,GDIPlus}FontData`. If we don't really need this field in `wxFont`, it would seem to be better to store it there, where it's actually needed.

OT3H, if we do store it in `wxFont`, finally, perhaps we could simply `WXAdjustToPPI()` and related code, which was added, IIRC, just to avoid storing the PPI in it initially? If any such simplifications were possible, I'd be in favour of doing what this PR does, but if the PPI is only going to be used in `wxGC` code, I'd rather store it in `wxGC`-specific objects, this just seems more logical.

MaartenBent commented:
I moved the DPI information to `wx{D2D,GDIPlus}FontData`. Public function `wxGraphicsContext ::CreateFont()` requires an extra argument for this (I forgot to document it, I noticed now). What do you think of this approach?

> For the `wxStyledTextCtrl` files, it's necessary to modify stc.h.in and stc.cpp.in

Thanks, totally forgot about this.

vadz commented on include/wx/graphics.h:
```diff
@@ -1033,7 +1033,7 @@ class WXDLLIMPEXP_CORE wxGraphicsRenderer : public wxObject
 
     // sets the font
-    virtual wxGraphicsFont CreateFont( const wxFont &font , const wxColour &col = *wxBLACK ) = 0;
+    virtual wxGraphicsFont CreateFont( const wxFont &font , const wxColour &col = *wxBLACK, const wxSize &dpi = wxSize() ) = 0;
```
Seeing that we operate with doubles everywhere in this code, maybe `dpi` argument should use `wxRealPoint` instead?

vadz commented on src/common/graphcmn.cpp:
```diff
@@ -944,16 +944,18 @@ wxGraphicsContext::CreateRadialGradientBrush(
-    return GetRenderer()->CreateFont(font,col);
+    wxDouble dpiX, dpiY;
+    GetDPI(&dpiX, &dpiY);
+    return GetRenderer()->CreateFont(font, col, wxSize(dpiX, dpiY));
```
If we keep using `wxSize()`, I'd add `wxRound()` calls here, even if it doesn't change much in practice.

vadz commented on src/common/graphcmn.cpp:
```diff
@@ -631,8 +631,8 @@ void wxGraphicsContext::GetDPI( wxDouble* dpiX, wxDouble* dpiY)
-        *dpiX = 72.0;
-        *dpiY = 72.0;
+        *dpiX = 96.0;
+        *dpiY = 96.0;
```
Are you sure about this? I think it's used for Cairo too and it definitely uses 72 and not 96.

MaartenBent replied:
Yes, I definitely want 96 here on Windows. This is also what [D2D ](https://docs.microsoft.com/en-us/windows/win32/api/d2d1/nf-d2d1-id2d1rendertarget-getdpi) and [GDI+](https://docs.microsoft.com/en-us/windows/win32/api/gdiplusgraphics/nf-gdiplusgraphics-graphics-getdpix) return. I'll investigate Cairo.

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -4572,14 +4568,6 @@ void wxD2DContext::Flush()
-    if (dpiX != NULL) *dpiX = x;
-    if (dpiY != NULL) *dpiY = y;
-}
-
```
Has this been removed because it's the same as `wxWindow::GetDPI()` used by the base class?

MaartenBent replied:
The base version uses the window, which is DPI-Aware. This call always returns the system-DPI.

Maybe I can do something like:
```c++
// call base version that uses window
wxGraphicsContext::GetDPI(...)
if (x == 0 || y ==0) { /* renderer specific version */) }
```
This way Cairo can also return 72.

vadz replied:
Yes, returning hardcoded 72 never sat well with me neither. But OTOH this is another one of not perfectly backwards compatible changes I grew to be wary about... So maybe we should rather add `bool TryGetDPI(x, y)` and keep returning 72 from `GetDPI()` by default?

vadz reviewed (approved):
I indeed prefer this version, thanks for the update! AFAICS this should be ready to merge.

Have you been able to check if the font size used by wxDC and wxGC for the same point value is actually the same in different DPIs by chance?

MaartenBent commented:
> I indeed prefer this version, thanks for the update! AFAICS this should be ready to merge.

I noticed some more issues. Because `CreateFont` is a public function, users can call it without a DPI argument and the default `wxSize()` will be used, resulting on font height `0`. In this case I have to initialize the DPI in the implementation.

> Have you been able to check if the font size used by wxDC and wxGC for the same point value is actually the same in different DPIs by chance?

I'm testing the drawing sample at different DPI. All three (`wxDC`, `GDI+` and `D2D`) have the same font sizes.

vadz commented:
> > I indeed prefer this version, thanks for the update! AFAICS this should be ready to merge.
> 
> I noticed some more issues. Because `CreateFont` is a public function, users can call it without a DPI argument and the default `wxSize()` will be used, resulting on font height `0`. In this case I have to initialize the DPI in the implementation.

In fact, I was thinking about this change not being 200% backwards-compatible because if the user code overrides `CreateFont()` (e.g. because they have some kind of a wrapper/faÃ§ade for wxGC), it would stop working. I decided not to mention this initially, but adding another `CreateFont()` overload, with 3 non-optional parameters, or even a new `CreateFontAtDPI()`, would be perfectly compatible _and_ might be actually simpler to implement, as you wouldn't have to initialize the DPI manually, so perhaps we should do it instead?

> > Have you been able to check if the font size used by wxDC and wxGC for the same point value is actually the same in different DPIs by chance?
> 
> I'm testing the drawing sample at different DPI. All three (`wxDC`, `GDI+` and `D2D`) have the same font sizes.

Excellent, thanks a lot!

MaartenBent committed:
Improve wxD2DContext::GetDPI and wxGDIPlusContext::GetDPI

If there is a valid wxWindow, use its DPI. Otherwise use a dedicated function
of the context to get the DPI. Don't use the common wxGraphicsContext::GetDPI
because this will return hard-coded 72 when there is no valid wxWindow.

MaartenBent commented:
I added `CreateFontAtDPI()` to`wxGraphicsContext` and improved `GetDPI()` of `wxD2DContext` and added `GetDPI()` to `wxGDIPlusContext`. This way it should all be backwards-compatible and not affect other renderers.

vadz commented:
Thanks! I'll look at this a.s.a.p., but I still didn't have time to actually test it. Do you think you've already done it sufficiently or should I still do it too before merging?

vadz commented on src/msw/graphics.cpp:
```diff
@@ -2387,30 +2388,24 @@ void wxGDIPlusContext::GetSize( wxDouble* width, wxDouble *height )
-    context->SetPageScale(dpiRatio);
+    if ( GetWindow() )
+    {
+        wxSize dpi = GetWindow()->GetDPI();
```
Micro suggestion
```suggestion
        const wxSize dpi = GetWindow()->GetDPI();
```

vadz reviewed (approved):
Thanks once again and I'm ready to merge it as soon as you give your green light (and maybe after testing it at least a bit myself).

BTW, if anybody else using actual high DPI monitors with Windows 10 (I just emulate them in a VM) could please test it, it would be definitely very welcome. TIA!

MaartenBent commented on src/msw/graphics.cpp:
```diff
@@ -1157,27 +1154,31 @@ wxGDIPlusFontData::wxGDIPlusFontData( wxGraphicsRenderer* renderer,
         style |= FontStyleBold;
 
-    Init(font.GetFaceName(), font.GetFractionalPointSize(), style, col);
+    REAL fontSize = (REAL)(!dpi.y
```
The `wxGraphicsRenderer` (with `CreateFont`) has no way of accessing the `wxGraphicsContext` (with the window/dpi functions).

vadz replied:
Ah, yes, of course, thanks.

MaartenBent committed:
Remove wxGDIPlusPrintingContext::SetFont, override DPI instead

MaartenBent commented:
Addressed the last comments. I'm OK with merging. Hopefully Travis does not act up...

I prefer that you test it before merging, just to be sure it actually behaves as you expect it to.

vadz commented:
OK, I'll try to test it a.s.a.p. In addition to run-time behaviour, I'd also like to check if all this exciting new stuff works with MinGW (at least the latest/recent versions of MinGW-w64, I understand that it won't work with some hoary MinGW-w32 from 2005). Would you know how to build wx apps using MinGW so that they use the correct manifest for per-monitor v2 DPI support by chance?

MaartenBent commented:
Easiest is to modify `include/wx/msw/amd64.manifest` and add the required items (see #334).

vadz commented:
Maybe we could have this available in the manifest but just conditional on some `wxUSE_DPI_AWARE_MANIFEST` or something like this, so that you could simply predefine this symbol to enable it? It should be much simpler than hunting for "the required items" (I'm not sure where does #334 list them, I just keep `hidpi.manifest` locally for my own use).

BTW, should we close #334 now?

MaartenBent commented:
The first post has a collapsed `<details></details>` section with a patch that has a dpi-aware manifest. #334 can now be closed, everything has been cherry-picked and merged. I'll have a look at adding `wxUSE_DPI_AWARE_MANIFEST`.

MaartenBent committed:
Fix wxSTC compilation without precompilead headers

Get and release the HDC, instead of using WindowHDC. Because stc.cpp is used on
all platforms, don't include wxMSW specific wx/msw/private.h.

Also use const int for the variables.

MaartenBent committed:
Improve button heights at high DPI on wxMSW

Don't use FromDIP together with ConvertDialogToPixels, because the font height
is already adjusted to the DPI.
Also limit the button height at higher DPI. Because the height determined by
ConvertDialogToPixels is higher than standard buttons use on Windows.

Closes #18528

vadz commented:
Thanks again, Maarten! I'm merging this, but I've applied more fixes on top of the last commit for the button size, see the discussion on wxTrac.
